op-batcher/batcher/driver.go

```diff
@@ -945,3 +919,136 @@ func logFields(xs ...any) (fs []any) {
 	}
 	return fs
 }
+
+type ChannelStatuser interface {
+	isFullySubmitted() bool
+	isTimedOut() bool
+	LatestL2() eth.BlockID
+	MaxInclusionBlock() uint64
+}
+
+type SyncActions struct {
+	clearState      *eth.BlockID
+	blocksToPrune   int
+	channelsToPrune int
+	waitForNodeSync bool
+	blocksToLoad    [2]uint64 // the range [start,end] that should be loaded into the local state.
+	// NOTE this range is inclusive on both ends, which is a change to previous behaviour.
+}
+
+func (s SyncActions) String() string {
+	return fmt.Sprintf(
+		"SyncActions{blocksToPrune: %d, channelsToPrune: %d, waitForNodeSync: %t, clearState: %v, blocksToLoad: [%d, %d]}", s.blocksToPrune, s.channelsToPrune, s.waitForNodeSync, s.clearState, s.blocksToLoad[0], s.blocksToLoad[1])
+}
+
+// computeSyncActions determines the actions that should be taken based on the inputs provided. The inputs are the current
+// state of the batcher (blocks and channels), the new sync status, and the previous current L1 block. The actions are returned
+// in a struct specifying the number of blocks to prune, the number of channels to prune, whether to wait for node sync, the block
+// range to load into the local state, and whether to clear the state entirely.
+func computeSyncActions[T ChannelStatuser](newSyncStatus eth.SyncStatus, prevCurrentL1 eth.L1BlockRef, blocks queue.Queue[*types.Block], channels []T, l log.Logger) SyncActions {
+
+	if newSyncStatus.HeadL1 == (eth.L1BlockRef{}) {
+		s := SyncActions{waitForNodeSync: true}
+		l.Warn("empty sync status", "syncActions", s)
+		return s
+	}
+
+	if newSyncStatus.CurrentL1.Number < prevCurrentL1.Number {
+		// This can happen when the sequencer restarts
+		s := SyncActions{waitForNodeSync: true}
+		l.Warn("sequencer currentL1 reversed", "syncActions", s)
+		return s
+	}
+
+	oldestBlock, hasBlocks := blocks.Peek()
+
+	if !hasBlocks {
+		s := SyncActions{
+			blocksToLoad: [2]uint64{newSyncStatus.SafeL2.Number + 1, newSyncStatus.UnsafeL2.Number},
+		}
+		l.Info("no blocks in state", "syncActions", s)
+		return s
+	}
+
+	if oldestBlock.NumberU64() > newSyncStatus.SafeL2.Number+1 {
+		s := SyncActions{
+			clearState:   &newSyncStatus.SafeL2.L1Origin,
+			blocksToLoad: [2]uint64{newSyncStatus.SafeL2.Number + 1, newSyncStatus.UnsafeL2.Number},
+		}
+		l.Warn("new safe head is behind oldest block in state", "syncActions", s)
+		return s
+	}
+
+	newestBlock := blocks[blocks.Len()-1]
+	numBlocksToDequeue := newSyncStatus.SafeL2.Number + 1 - oldestBlock.NumberU64()
+
+	if numBlocksToDequeue > uint64(blocks.Len()) {
+		// This could happen if the batcher restarted.
+		// The sequencer may have derived the safe chain
+		// from channels sent by a previous batcher instance.
+		s := SyncActions{
+			clearState:   &newSyncStatus.SafeL2.L1Origin,
```

Comment:
What's the correct value for this? Can we update the readme to describe how this mechanism works (e.g. to try and keep. a steady pulse of batcher transactions). 

Is `waitNodeSyncAndClearState()` now dead code? Can we remove it?

Reply:
`waitNodeSyncAndClearState()` is still used for when we try to get blocks from the sequencer. That can't be encapsulated into the new pure function, but can still indicate a reorg.

Reply:
Updated the readme here [ba62afe](https://github.com/ethereum-optimism/optimism/pull/13060/commits/ba62afe93b5e0687cea5cdd661aecae8563479b2)